File: src/services/api/reader.ts
import { invoke } from '@tauri-apps/api/core'
import type { Chapter, ReadingProgress, ReadingRecord, ReadingStats, SearchContentResponse } from './types'

export interface Bookmark {
  book_id: number
  chapter_id: number
  offset: number
  mode: 'scroll' | 'paged'
}

/** A bookmark joined with the book and chapter it points at (书签 page). */
export interface BookmarkEntry extends Bookmark {
  book_title: string
  book_author: string | null
  chapter_title: string
  chapter_number: number
  updated_at: string
}

/** Every bookmark in the library, newest first (`list_bookmarks`). */
export function listBookmarks(): Promise<BookmarkEntry[]> {
  return invoke<BookmarkEntry[]>('list_bookmarks')
}

export function getBookmark(bookId: number, chapterId: number): Promise<Bookmark | null> {
  return invoke('get_bookmark', { bookId, chapterId })
}

export function saveBookmark(
  bookId: number,
  chapterId: number,
  offset: number,
  mode: 'scroll' | 'paged',
): Promise<void> {
  return invoke('save_bookmark', { bookId, chapterId, offset, mode })
}

export function deleteBookmark(bookId: number, chapterId: number): Promise<void> {
  return invoke('delete_bookmark', { bookId, chapterId })
}

export function listChapters(bookId: number): Promise<Chapter[]> {
  return invoke<Chapter[]>('list_chapters', { bookId })
}

export function readChapter(chapterId: number): Promise<Chapter> {
  return invoke('read_chapter', { chapterId })
}

/**
 * Warms the chapters around the open one, the way the reference app's
 * `ReadBook.preDownload()` does. Fire and forget: the backend answers
 * immediately and downloads in the background.
 */
export function prefetchChapters(bookId: number, chapterId: number): Promise<void> {
  return invoke<void>('prefetch_chapters', { bookId, chapterId })
}

/** Stops the prefetch window still running for this book (closing the reader). */
export function cancelPrefetch(bookId: number): Promise<void> {
  return invoke<void>('cancel_prefetch', { bookId })
}

/** Chapters prefetched ahead of the open one; `0` turns prefetching off. */
export function getReaderPrefetchNum(): Promise<number> {
  return invoke<number>('get_reader_prefetch_num')
}

export function setReaderPrefetchNum(chapters: number): Promise<number> {
  return invoke<number>('set_reader_prefetch_num', { chapters })
}

/** Re-fetches the catalog from the book's source and returns the merged list. */
export function refreshCatalog(bookId: number): Promise<Chapter[]> {
  return invoke<Chapter[]>('refresh_catalog', { bookId })
}

/**
 * 正文搜索：scans every chapter of the book whose body is available offline.
 * Resolves once with the full hit list (progress arrives on the
 * `search-content-progress` event), so the reader never has to poll.
 */
export function searchBookContent(bookId: number, query: string, regex = false): Promise<SearchContentResponse> {
  return invoke<SearchContentResponse>('search_book_content', { bookId, query, regex })
}

/** Stops the scan in flight for this book; safe to call when none is running. */
export function cancelBookContentSearch(bookId: number): Promise<void> {
  return invoke<void>('cancel_book_content_search', { bookId })
}

export function fetchOnlineContent(sourceId: number, chapterUrl: string, chapterId?: number): Promise<string> {
  return invoke<string>('fetch_online_content', { sourceId, chapterUrl, chapterId })
}

export function getReadingProgress(bookId: number): Promise<ReadingProgress | null> {
  return invoke<ReadingProgress | null>('get_reading_progress', { bookId })
}

export function saveReadingProgress(
  bookId: number,
  chapterId: number,
  offset: number,
  anchorIndex?: number,
  anchorRatio?: number,
): Promise<void> {
  return invoke<void>('save_reading_progress', { bookId, chapterId, offset, anchorIndex, anchorRatio })
}

export function getReadingRecord(bookId: number): Promise<ReadingRecord | null> {
  return invoke<ReadingRecord | null>('get_reading_record', { bookId })
}

export function addReadingTime(bookId: number, durationSeconds: number): Promise<void> {
  return invoke<void>('add_reading_time', { bookId, durationSeconds })
}

export function getReadingStats(): Promise<ReadingStats> {
  return invoke<ReadingStats>('get_reading_stats')
}

export function setReadingGoal(minutes: number): Promise<ReadingStats> {
  return invoke<ReadingStats>('set_reading_goal', { minutes })
}

File: src-tauri/src/command/reader.rs
use crate::{
    app::AppState,
    domain::{Chapter, ReadingProgress, ReadingRecord, ReadingStats},
    error::AppError,
    service::{reader_service::ReaderService, settings_service::SettingsService},
};
use tauri::{AppHandle, Emitter, State};

#[tauri::command(rename = "read_chapter")]
pub async fn read_chapter_cmd(
    state: State<'_, AppState>,
    chapter_id: i64,
) -> Result<Chapter, AppError> {
    ReaderService::new(state.database()?)
        .read_chapter(chapter_id)
        .await
}

/// Warms the chapters around the open one (`legado` `ReadBook.preDownload`).
/// Deliberately fire-and-forget: reading must never wait on a prefetch, and a
/// book without a source simply has nothing to download.
#[tauri::command(rename = "prefetch_chapters")]
pub async fn prefetch_chapters_cmd(
    state: State<'_, AppState>,
    book_id: i64,
    chapter_id: i64,
) -> Result<(), AppError> {
    let pool = state.database()?;
    tauri::async_runtime::spawn(async move {
        if let Err(error) = ReaderService::new(pool)
            .prefetch_around(book_id, chapter_id)
            .await
        {
            tracing::warn!(target: "reader", book_id, chapter_id, %error, "chapter prefetch aborted");
        }
    });
    Ok(())
}

/// Stops the window in flight when the reader closes the book.
#[tauri::command(rename = "cancel_prefetch")]
pub async fn cancel_prefetch_cmd(book_id: i64) -> Result<(), AppError> {
    crate::service::reader_service::cancel_prefetch(book_id)
}

#[tauri::command(rename = "get_reader_prefetch_num")]
pub async fn get_reader_prefetch_num_cmd(state: State<'_, AppState>) -> Result<i64, AppError> {
    SettingsService::new(state.database()?)
        .reader_prefetch_num()
        .await
}

#[tauri::command(rename = "set_reader_prefetch_num")]
pub async fn set_reader_prefetch_num_cmd(
    state: State<'_, AppState>,
    chapters: i64,
) -> Result<i64, AppError> {
    let settings = SettingsService::new(state.database()?);
    settings.save_reader_prefetch_num(chapters).await?;
    settings.reader_prefetch_num().await
}

#[tauri::command(rename = "list_chapters")]
pub async fn list_chapters_cmd(
    state: State<'_, AppState>,
    book_id: i64,
) -> Result<Vec<Chapter>, AppError> {
    ReaderService::new(state.database()?)
        .list_chapters(book_id)
        .await
}

#[tauri::command(rename = "refresh_catalog")]
pub async fn refresh_catalog_cmd(
    app: AppHandle,
    state: State<'_, AppState>,
    book_id: i64,
) -> Result<Vec<Chapter>, AppError> {
    let before = ReaderService::new(state.database()?)
        .list_chapters(book_id)
        .await?;
    let chapters = ReaderService::new(state.database()?)
        .refresh_catalog(book_id)
        .await?;
    let known = before
        .into_iter()
        .map(|chapter| chapter.number)
        .collect::<std::collections::HashSet<_>>();
    let added = chapters
        .iter()
        .filter(|chapter| !known.contains(&chapter.number))
        .count();
    app.emit(
        "chapter-updated",
        serde_json::json!({ "book_id": book_id, "count": chapters.len(), "added": added }),
    )
    .map_err(AppError::database)?;
    Ok(chapters)
}

#[tauri::command(rename = "fetch_online_content")]
pub async fn fetch_online_content_cmd(
    state: State<'_, AppState>,
    source_id: i64,
    chapter_url: String,
    chapter_id: Option<i64>,
) -> Result<String, AppError> {
    ReaderService::new(state.database()?)
        .fetch_online_content(source_id, &chapter_url, chapter_id)
        .await
}

#[tauri::command(rename = "get_reading_progress")]
pub async fn get_reading_progress_cmd(
    state: State<'_, AppState>,
    book_id: i64,
) -> Result<Option<ReadingProgress>, AppError> {
    ReaderService::new(state.database()?)
        .progress(book_id)
        .await
}

#[tauri::command(rename = "save_reading_progress")]
pub async fn save_reading_progress_cmd(
    state: State<'_, AppState>,
    book_id: i64,
    chapter_id: i64,
    offset: i64,
    anchor_index: Option<i64>,
    anchor_ratio: Option<f64>,
) -> Result<(), AppError> {
    let service = ReaderService::new(state.database()?);
    if let (Some(anchor_index), Some(anchor_ratio)) = (anchor_index, anchor_ratio) {
        service
            .save_progress_anchor(book_id, chapter_id, offset, anchor_index, anchor_ratio)
            .await
    } else {
        service.save_progress(book_id, chapter_id, offset).await
    }
}

#[tauri::command(rename = "get_reading_record")]
pub async fn get_reading_record_cmd(
    state: State<'_, AppState>,
    book_id: i64,
) -> Result<Option<ReadingRecord>, AppError> {
    ReaderService::new(state.database()?)
        .reading_record(book_id)
        .await
}

#[tauri::command(rename = "add_reading_time")]
pub async fn add_reading_time_cmd(
    state: State<'_, AppState>,
    book_id: i64,
    duration_seconds: i64,
) -> Result<(), AppError> {
    ReaderService::new(state.database()?)
        .add_reading_time(book_id, duration_seconds)
        .await
}

#[tauri::command(rename = "get_reading_stats")]
pub async fn get_reading_stats_cmd(
    state: State<'_, AppState>,
) -> Result<ReadingStats, AppError> {
    ReaderService::new(state.database()?).reading_stats().await
}

#[tauri::command(rename = "set_reading_goal")]
pub async fn set_reading_goal_cmd(
    state: State<'_, AppState>,
    minutes: i64,
) -> Result<ReadingStats, AppError> {
    ReaderService::new(state.database()?)
        .set_reading_goal(minutes)
        .await
}
